swarm/src/lib.rs

```diff
@@ -1800,7 +1800,7 @@ impl NetworkInfo {
 ///
 /// If the given address is not yet a `p2p` address for the given peer,
 /// the `/p2p/<peer-id>` protocol is appended to the returned address.
-fn p2p_addr(peer: Option<PeerId>, addr: Multiaddr) -> Result<Multiaddr, Multiaddr> {
+pub fn p2p_addr(peer: Option<PeerId>, addr: Multiaddr) -> Result<Multiaddr, Multiaddr> {
```

Comment:
Let's not make this public please. Such utility functions are not worth it to be in the public API.

Reply:
Plus, on all new usage sites, the error that is produced here can never happen, meaning this function is actually not a good fit for those usecases.

Reply:
No problem. I did such a thing because a common utility method was necessary in several different crates (`libp2p-identify`, `libp2p-kad` and `libp2p-mdns`) and I did not want to duplicate code. Do you have a better location in mind ? Maybe the best location would be a public method on `multiaddr` itself (but it is in a different repository).

Is it completely sure that it can never have a `/p2p` suffix in these `multiaddr` ? I though it was a good security precaution to be able to handle an already present and different `peer_id`.

On all regards, I think it could be useful to end users to have access to such a method (it is actually the case for us). Indeed, other users could rely on equality of `multiaddr` and experience side effects, the same as we do currently, if they do not handle well the presence or absence of `/p2p` suffix. If we don't want to return a result, we could also just override the present `peer_id` if it differs from the provided one.

Reply:
> Maybe the best location would be a public method on `multiaddr` itself (but it is in a different repository).

Yeah I guess it is such a common usecase, it might be worth adding a utility function on `Multiaddr`. Mind proposing a patch at https://github.com/multiformats/rust-multiaddr?

Reply:
As proposed by @mxinden on https://github.com/multiformats/rust-multiaddr/pull/102, this is the proposal using the `with_p2p` recently added on `Multiaddr`. 

The following part in `Cargo.toml` will of course be replaced when `Multiaddr:0.18.1` will be released : 
```toml
# multiaddr = "0.18.0"
multiaddr = { git = "https://github.com/multiformats/rust-multiaddr", rev = "d0e9ec219645ce8bd8ae7926e4b7316a67c35f84" }
```